### ISSUE TITLE
Large amounts of coroutine recursion causes a crash

### ISSUE THREAD

**Issue by MCJack123** [MCJack123/craftos2#124]
**Describe the bug**
When resuming a high number of stacked coroutines, CraftOS-PC will crash. This is due to a stack overflow.

**To Reproduce**
1. Open a Lua prompt
2. Paste this code: `f = function(n) print(n) coroutine.resume(coroutine.create(f), n+1) end f(1)`
3. Observe crash

**Expected behavior**
The coroutines are able to continue executing infinitely, or error after some number, without crashing.

**Screenshots**
N/A. This typically causes an illegal instruction, or a segmentation fault somewhere. On Mac it can recurse ~500 levels deep before crashing.

**Environment (please complete the following information):**
 - OS: Any
 - Version: v2.3.4 (Any)
 - Compiled from source? N/A

**Additional context**
CCEmuX supports up to 89640 levels before erroring out, but this may be due to the 7-second timeout.

**Comment by SquidDev** [comment]
>  CCEmuX supports up to 89640 levels before erroring out, but this may be due to the 7-second timeout.

This will be due to the timeout - Cobalt effectively unwinds the Java stack before switching into another coroutine, so its depth is unlimited.

[PUC Lua](https://www.lua.org/source/5.1/ldo.c.html#lua_resume) treats coroutine resumes as a "C call", which gives an effective max-depth 200. This is much lower than the above error, but means it throws a "C Stack Overflow" error rather than crashing.

Your version of Lua appears to be missing these lines from `lua_resume`. The implementation is very different to PUC's, so I assume you're using one of the coroutine patches?

```lua
if (L->nCcalls >= LUAI_MAXCCALLS)
    return resume_error(L, "C stack overflow");
```

**Comment by MCJack123** [comment]
Indeed, though while initially debugging the crash I found that `nCcalls` is not updated on resume, and was still much lower than the resume count. There are a few things I need to add; this issue is mainly so I don't forget to do it tomorrow :P.

I'm using a patch from [lua-users](http://lua-users.org/wiki/ResumableVmPatch) to allow yields across C calls. I guess this patch must have removed the stack check in some way, unless the check was added between 5.1.0 and 5.1.5 and didn't make it through the three-way merge (the original patch is for 5.1, so it had to be upgraded for 5.1.5).

**Comment by SquidDev** [comment]
Ahh right, I forget how different the patched version of 5.1 is compared with 5.3. Disregard what I said above, sorry!

I don't know if it's worth experimenting with the [Coco](https://coco.luajit.org/) patch instead. I believe it has the same roots (developed by Mike Pall), but uses a slightly different approach and is a little more up-to-date.
